File: leetcode/5.最长回文子串.rs
/*
 * @lc app=leetcode.cn id=5 lang=rust
 *
 * [5] 最长回文子串
 */

// @lc code=start
impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        let arr: Vec<char> = s.chars().collect();
        let (mut start_idx, mut max) = (0, 1);
        for i in 0..arr.len() {
            let (start1, len1) = Solution::expand_center_around(&arr, i, i);
            let (start2, len2) = Solution::expand_center_around(&arr, i, i + 1);
            if len2 > len1 {
                if len2 > max {
                    start_idx = start2;
                    max = len2;
                }
            } else {
                if len1 > max {
                    start_idx = start1;
                    max = len1;
                }
            }
        }
        arr[start_idx..start_idx+max].iter().collect(
    }
    fn expand_center_around(s: &Vec<char>, mut lt: usize, mut rt: usize) -> (usize, usize) {
        while rt < s.len() && s[lt] == s[rt] {
            if lt == 0 {
                return (0, rt + 1);
            }
            lt -= 1;
            rt += 1;
        }
        (lt + 1, rt - lt - 1)
    }
}
// @lc code=end

File: leetcode/5.最长回文子串.go
/*
 * @lc app=leetcode.cn id=5 lang=golang
 *
 * [5] 最长回文子串
 */

// @lc code=start
func longestPalindrome(s string) string {
	start, end := 0, 0
	for i := 0; i < len(s); i++ {
		left1, right1 := ExpandCenterAround(s, i, i)
		left2, right2 := ExpandCenterAround(s, i, i + 1)
		if right1-left1>end-start {
			start, end = left1, right1
		}
		if right2-left2>end-start {
			start, end = left2, right2
		}
	}
	return s[start:end+1]
}
func ExpandCenterAround(s string, left, right int) (int, int) {
	for ; left >= 0 && right < len(s) && s[left] == s[right]; left, right = left-1, right+1 {
	}
	return left + 1, right - 1
}

// @lc code=end

